### DISCUSSION TITLE
Allow doctrine to work correctly with counters

### DISCUSSION THREAD

**Discussion by awwar** [doctrine/doctrine2#11043] [Support Questions] [answered]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

First of all, I want to say thank you to your team. Doctrine allows complex things to be made very simple, understandable and supportable.

But there are certain things that still have to be processed with raw queries.
For example, increasing counters:

```php
#[Entity]
class Article
{
    #[Id, Column(type: 'integer'), GeneratedValue]
    private int $id;

    #[AsCounter] // <------ An attribute that marks a field as a counter
    #[Column(type: 'integer')]
    private int $views;

    #[Column(type: 'string')]
    private string $authorName;

    public function addViews(int $views): void
    {
        $this->views += $views;
    }

    public function getViews(): int
    {
        return $this->views;
    }

    public function setAuthorName(string $newName): void
    {
        $this->authorName = $newName;
    }
}

$article = $this->em->find('Article', 1);

echo $article->getViews(); // 13

$article->addViews(4);
$article->setAuthorName("Jon");

$this->em->flush();
```

What happens after `flush` now, without `AsCounter` attribute:

```sql
UPDATE articles SET author_name = 'Jon', views = 17  WHERE id = 1
                                                ^^^
```

But it would be cool if, thanks to the `AsCounter` attribute, it worked like this:

```sql
UPDATE articles SET author_name = 'Jon', views = views + 4 WHERE id = 1
                                                 ^^^^^^^^^
```

**Comment by derrabus** [comment]
Thank you for your proposal. I'm afraid this feature is a bit too specific to include it in the ORM library.

**Comment by awwar** [comment]
@derrabus Thanks for the answer. Is there any extension point to add this functionality to my project?

I know that i can add a custom type:

```php
class CounterType extends Type
{
    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform) { /* ToDO */ }

    public function convertToPHPValue($value, AbstractPlatform $platform) { /* ToDO */ }

    public function convertToDatabaseValue(int $value, AbstractPlatform $platform) 
    { 
         return sprintf('%s %d', $columnName, $value);
    }

    public function getName()
    {
        return 'counter'
    }
}
```

But I don’t know how to specify the `$columnName` in the `convertToDatabaseValue` method.

**Or is this not the right way and should be done differently?**

**Reply by derrabus** [reply]
I'm moving this issue to support.

Using a custom type is clearly a hack. I would strongly advise against it.

Maybe someone else has a better idea, but I'm not aware that we have a good extension point that you could leverage here. You should however be able to build a DQL UPDATE statement for this.

If your concern is a lost update in concurrent requests, you could also lock the entity when selecting it.

**Comment by beberlei** [comment, marked as answer]
You can use a second property with the additional value, a post update+persist listener and unitofwork scheduleExtraUpdate API or directly issue a second update at that point.

**Reply by awwar** [reply]
Thanks for the ideas. I'll try to use them.
